src/View/InstallQuickInput.ts

```diff
@@ -18,142 +18,305 @@ import * as vscode from 'vscode';
-    return undefined;
+
+    this.version = state.selectedItem.label;
+    this.toolchain = toolchains[versions.indexOf(this.version)];
+
+    this.changeCurrentStepAfter(this.pickVersion.name, state as State);
   }
 
-  async function updateBackend(input: MultiStepInput, state: Partial<State>) {
-    if (state.toolchainEnv === undefined || state.toolchainType === undefined) {
+  async updateBackend(input: MultiStepInput, state: Partial<State>) {
+    if (this.toolchainEnv === undefined || this.toolchainType === undefined) {
+      this.error = 'toolchainenv is undefined.';
       throw Error('toolchainenv is undefined.');
     }
 
-    await state.toolchainEnv.prerequisites();
+    await this.toolchainEnv.prerequisites();
 
     // NOTE(jyoung)
     // Prerequisites request shows the password quick input and this input is
     // automatically added to MultiStepInput's steps. If the user clicks
     // 'back' button on quick input, the previous step is shown, and the password
     // input is show at this time, so it must be removed directly from the steps.
     input.steps.pop();
-    return pickVersion(input, state);
+
+    this.isUpdateBackend = false;
   }
 
-  function shouldResume() {
-    // Could show a notification with the option to resume.
-    return new Promise<boolean>(
-        (resolve, reject) => {
-            // noop
-        });
+  getMultiSteps(state: Partial<State>) {
+    if (state.current === undefined || state.current === InstallQuickInputStep.pickVersion) {
+      throw Error('state is wrong: ' + String(state.current));
+    }
+
+    // pickBackend -> pickType -> pickVersion
+    const steps: InputStep[] = [
+      (input => this.pickBackend(input, state)), (input => this.pickType(input, state)),
+      (input => this.pickVersion(input, state))
+    ];
+    // stepPickBackend = steps.slice(0, 1);
+    // stepPickType = steps.slice(0, 2);
+    // stepPickVersion = steps.slice(0, 3);
+    return steps.slice(0, state.current + 1);
   }
 
-  const state = await collectInputs();
-  return new Promise<[ToolchainEnv, Toolchain]>((resolve, reject) => {
-    if (state.toolchainEnv && state.toolchain) {
-      resolve([state.toolchainEnv, state.toolchain]);
-    } else {
-      reject();
+  async collectInputs() {
+    const state = {selectedItem: undefined, current: InstallQuickInputStep.unset} as Partial<State>;
+    while (state.current! < InstallQuickInputStep.pickVersion) {
+      await MultiStepInput.runSteps(this.getMultiSteps(state));
+      if (this.isUpdateBackend) {
+        await MultiStepInput.run(input => this.updateBackend(input, state));
+      }
     }
-  });
+  }
 }
+
+async function showInstallQuickInput(): Promise<[ToolchainEnv, Toolchain]> {
+  const quickInput = new InstallQuickInput();
+  await quickInput.collectInputs();
+
+  const error = quickInput.getError();
+  if (error !== undefined) {
+    vscode.window.showErrorMessage(error);
+    throw Error(error);
+  }
```

Comment:
Q) (it can be a really elementary question, but) AFAIU, any cases whenever `InstallQuickInput.error` set, `Throw(error)` has been called. Without try-catch, can the code approaches to line 313?

Reply:
Until now, your saying is right. The code flow cannot reach there. In the future, however, another new code would be introduced and this could happen. It's a kind of code to protect against unexpected cases.